libraries/AP_NMEA_Output/AP_NMEA_Output.cpp

```diff
@@ -0,0 +1,174 @@
+            break;
+        }
+        _num_outputs++;
+    }
+}
+
+AP_NMEA_Output* AP_NMEA_Output::probe()
+{
+    if (!_singleton && AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_NMEAOutput, 0) != nullptr) {
+       _singleton = new AP_NMEA_Output();
+    }
+
+    return _singleton;
+}
+
+uint8_t AP_NMEA_Output::_nmea_checksum(const char *str)
+{
+    uint8_t checksum = 0;
+    const uint8_t* bytes = (const uint8_t*) str;
+
+    for (uint16_t i = 1; str[i]; i++) {
+        checksum ^= bytes[i];
+    }
+
+    return checksum;
+}
+
+void AP_NMEA_Output::update()
+{
+    uint16_t now = AP_HAL::millis16();
+    uint16_t time_diff = now - _last_sent;
+
+    // only send at 10Hz
+    if (time_diff < 100) {
+        return;
+    }
+
+    // get time and date
+    uint64_t time_usec;
+    bool time_valid = AP::rtc().get_utc_usec(time_usec);
+
+    if (!time_valid) {
+        return;
+    }
+
+    // not completely accurate, our time includes leap seconds and time_t should be without
+    time_t time_sec = time_usec / 1000000;
+    struct tm* tm = gmtime(&time_sec);
+
+    // format time string
+    char tstring[11];
+    snprintf(tstring, sizeof(tstring), "%02u%02u%06.3f", tm->tm_hour, tm->tm_min, tm->tm_sec + (time_usec % 1000000) * 1.0e-6);
+
+    // format date string
+    char dstring[7];
+    snprintf(dstring, sizeof(dstring), "%02u%02u%02u", tm->tm_mday, tm->tm_mon+1, tm->tm_year % 100);
+
+    AP_AHRS_NavEKF& ahrs = AP::ahrs_navekf();
+
+    // get location (note: get_position from AHRS always returns true after having GPS position once)
+    Location loc{};
+    bool pos_valid = ahrs.get_location(loc);
+
+    // format latitude
+    char lat_string[13];
+    double deg = fabs(loc.lat * 1.0e-7f);
+    snprintf(lat_string,
+             sizeof(lat_string),
+             "%02u%08.5f,%c",
+             (unsigned) deg,
+             (deg - int(deg)) * 60,
+             loc.lat < 0 ? 'S' : 'N');
+
+    // format longitude
+    char lng_string[14];
+    deg = fabs(loc.lng * 1.0e-7f);
+    snprintf(lng_string,
+             sizeof(lng_string),
+             "%03u%08.5f,%c",
+             (unsigned) deg,
+             (deg - int(deg)) * 60,
+             loc.lng < 0 ? 'W' : 'E');
+
+    // format GGA message
+    char* gga = nullptr;
+    int16_t gga_res = asprintf(&gga,
+                               "$GPGGA,%s,%s,%s,%01d,%02d,%04.1f,%07.2f,M,0.0,M,,",
+                               tstring,
+                               lat_string,
+                               lng_string,
+                               pos_valid ? 1 : 0,
+                               pos_valid ? 6 : 3,
+                               2.0,
+                               loc.alt / 100.0f);
+    char gga_end[6];
+    snprintf(gga_end, sizeof(gga_end), "*%02X\r\n", (unsigned) _nmea_checksum(gga));
+
+    // get speed
+    Vector2f speed = ahrs.groundspeed_vector();
+    float speed_knots = norm(speed.x, speed.y) * 1.94384449f;
```

Comment:
Magic numbers need to be a define with a name and/or commented well.

Reply:
Tell that to sitl_gps.cpp 😛